CODEBASE.md

```diff
@@ -62,13 +42,60 @@ These ops is being upstreamed to TF (with some minor differences) and we will mo
+We offer mixed precision training, which substantially improves throughput on NVIDIA tensor cores. It is important to note that you need to both pass in the `--fp16` flag to `train.py` AND set the `TENSORPACK_FP16` environmental variable to 1. `--fp16` tells the model to use FP16, while the `TENSORPACK_FP16` envvar enables loss scaling.
 
 Loss scaling occurs inside of the Tensorpack library, not in the MaskRCNN model code, which is why you need to set both. This should probably be addressed in the future.
 
+## NHWC convolution kernels
+
+According to Nvidia's update in MLPerf v0.6 - https://devblogs.nvidia.com/nvidia-boosts-ai-performance-mlperf-0-6/, tensor core accelerated convolution kernels expect NHWC (or “channels-last) layout. We added options to transpose convolution input data format to NHWC for backbone, FPN, RPN-head and mask-head.
+
+For tensorflow, NHWC kernel can only be activated by XLA or in Nvidia's NGC container. Since XLA doesn't work in our case (our tensor shapes are dynamic and tensorpack tower function can not be compiled with XLA), we use NGC container and export `TF_ENABLE_NHWC=1`
+
+## Disable Cudnn autotune and use aspect ratio grouping
+
+We disabled the cudnn autotune by export `TF_CUDNN_USE_AUTOTUNE=0` since the cudnn warmup is pretty slow and tuning result does not deliver a significant improve.
+
+When training with bs > 1, if the images sizes differ a lot in the same batch, padding will introduce extra computations which are not negligible. The idea is to batch the images of the same aspect ratio, i.e. height > width or height < width.
+
+Reference: https://github.com/facebookresearch/Detectron/blob/35fede2bd11176c600617e60ea4e7cc458cf4d64/detectron/roi_data/loader.py#L139-L157
+
+## Async evaluation and faster coco evaluation
+
+MLPerf v0.6 requires to evaluate every epoch and stop the training once the target accuracy is reached (box/mask 0.377/0.339). We start a background thread overlap the coco eval with the next iteration computation. Use the `--async_eval` to enable this feature.
+
+The coco eval usually takes more than 2 minutes. This will be a bottleneck for 24 nodes when each epoch takes less than 2 minutes. We adopt the [Nvidia's own implementation](https://github.com/NVIDIA/cocoapi/) of coco eval, which cut the time to less than 10 seconds.
+
+## Determinism
+
+Determinism can help to converge to a more stable result across different runs. It is also helpful for debugging purposes. By setting up op level seed for the TF graph and a generator for other random numbers, we have fully determinism for the forward computation. However for backprop, the ROIAlign op's CUDA kernel uses `atomicAdd` which introduces randomness.
+
```

Comment:
User should also set the TF_CUDNN_DETERMINISTIC flag

Reply:
> User should also set the TF_CUDNN_DETERMINISTIC flag, or turn off autotune

Good point, you can add it into the doc.

Reply:
Just mentioned that the TF_CUDNN_USE_AUTOTUNE flag should be 0, and is by default. That should be sufficient.